kubric/core.py

```diff
@@ -0,0 +1,212 @@
+# Copyright 2020 The Kubric Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import uuid
+from collections import defaultdict
+
+import traitlets as tl
+import mathutils
+from traitlets import default, validate
+
+import kubric.traits as kt
+from kubric.color import Color
+
+
+class Asset(tl.HasTraits):
+  uid = tl.Unicode(read_only=True)
+
+  def __init__(self, *args, **kwargs):
+    super().__init__(*args, **kwargs)
+    self.destruction_callbacks = []
+    self.keyframe_callbacks = []
+    self.keyframes = defaultdict(dict)
+
+  @default('uid')
+  def _uid(self):
+    return str(uuid.uuid4())
+
+  def keyframe_insert(self, member: str, frame: int):
+    if not self.has_trait(member):
+      raise KeyError('Unknown member "{}".'.format(member))
+    self.keyframes[member][frame] = getattr(self, member)
+    for func in self.keyframe_callbacks:
+      func(owner=self, member=member, frame=frame)
+
+  def __hash__(self):
+    return object.__hash__(self)
+
+  def __eq__(self, other):
+    return id(self) == id(other)
+
+  def __del__(self):
+    for func in self.destruction_callbacks:
+      func(owner=self)
+
+
+# ## ### ####  Materials  #### ### ## #
+
+class Material(Asset):
+  pass
+
+
+class PrincipledBSDFMaterial(Material):
+  color = kt.RGBA(default_value=Color.from_name('white'))
+  metallic = tl.Float(0.)
+  specular = tl.Float(0.5)
+  specular_tint = tl.Float(0.)
+  roughness = tl.Float(0.4)
+  ior = tl.Float(1.45)
+  transmission = tl.Float(0)
+  transmission_roughness = tl.Float(0)
+  emission = kt.RGBA(default_value=Color.from_name('black'))
+
+
+class MeshChromeMaterial(Material):
+  color = kt.RGBA(default_value=Color.from_name('white'))
+  roughness = tl.Float(0.4)
+
+# ## ### ####  3D Objects  #### ### ## #
```

Comment:
Does the google coding standard have a special comment for this type of separators?
So we avoid arguing w.r.t. personal choices :)

Reply:
I have no idea. This is a random choice on my part, which I do not feel strongly about.

Reply:
Ok, in that case I might globally update it later with something else ;)